scheduler/src/cook/mesos/sandbox.clj

```diff
@@ -265,29 +265,87 @@
 
               :else
               (recur remaining-hostnames pending-sync-agent-send-performed))))))
-    {:error-handler (fn sandbox-publisher-error-handler [ex]
+    {:error-handler (fn start-host-sandbox-syncer-error-handler [ex]
                       (log/error ex "Sync of sandbox directories on pending hosts failed"))}))
 
+(defn- remove-task-ids-with-sandbox
+  "Filters the task-ids which have a sandbox directory from task-id->sandbox."
+  [datomic-conn task-id->sandbox]
+  (letfn [(retrieve-task-ids-with-sandbox [datomic-db task-ids]
+            (d/q '[:find [?t ...]
+                   :in $ [?t ...]
+                   :where
+                   [?e :instance/task-id ?t]
+                   [?e :instance/sandbox-directory ?s]]
+                 datomic-db task-ids))
+          (remove-task-id-with-sandbox [task-id->sandbox task-id]
+            (dissoc! task-id->sandbox task-id))
+          (update-filter-entries-histogram! [task-id->sandbox]
+            (histograms/update! sandbox-updater-filter-entries (count task-id->sandbox))
+            task-id->sandbox)]
+    (timers/time!
+      sandbox-updater-filter-duration
+      (->> (keys task-id->sandbox)
+           (retrieve-task-ids-with-sandbox (d/db datomic-conn))
+           (reduce remove-task-id-with-sandbox (transient task-id->sandbox))
+           persistent!
```

Comment:
Why do we need `transient`/`persistent!` here?

Reply:
We used transient here as we can end up whittling down the map from a size of hundreds or thousands to tens.